packages/blade/src/components/FileUpload/FileUpload.web.tsx

```diff
@@ -0,0 +1,367 @@
+    isDisabled,
+    isRequired,
+    necessityIndicator,
+    fileList,
+    testID,
+    label,
+    labelPosition = 'top',
+    accessibilityLabel,
+    validationState,
+    helpText,
+    errorText,
+    maxCount,
+    maxSize,
+    ...styledProps
+  },
+  ref,
+): React.ReactElement => {
+  const { platform } = useTheme();
+  const [selectedFiles, setSelectedFiles] = useState<BladeFileList>(fileList ?? []);
+  const [errorMessage, setErrorMessage] = useState(errorText);
+  const [internalValidationState, setInternalValidationState] = useState('none');
+  const [isActive, setIsActive] = useState(false);
+
+  const isMultiple = uploadType === 'multiple';
+  const isOneFileSelectedWithSingleUpload = !isMultiple && selectedFiles.length === 1;
+  const inputLabelPosition = platform === 'onMobile' ? 'top' : labelPosition;
+  const isLabelLeftPositioned = inputLabelPosition === 'left';
+  const willRenderHintText = isOneFileSelectedWithSingleUpload
+    ? false
+    : Boolean(helpText) || Boolean(errorMessage);
+
+  const showError = validationState === 'error' || internalValidationState === 'error';
+  const showHelpText = !showError && helpText;
+  const accessibilityText =
+    accessibilityLabel ?? `,${showError ? errorText : ''} ${showHelpText ? helpText : ''}`;
+  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileuploadinput');
+
+  const accessibilityProps = makeAccessible({
+    required: Boolean(isRequired),
+    invalid: Boolean(showError),
+    disabled: Boolean(isDisabled),
+    describedBy: helpTextId,
+  });
+
+  // In control mode attach a unique id to each file if not provided
+  useMemo(() => {
+    for (const file of selectedFiles) {
+      if (!file.id) {
+        file.id = `${new Date().getTime().toString()}${Math.floor(Math.random() * 1000000)}`;
+      }
+    }
+  }, [selectedFiles]);
+
+  const handleFilesChange = useCallback((inputFiles: BladeFileList) => {
+    setSelectedFiles((prevFiles) => {
+      if (prevFiles.length > 0) {
+        const allFiles = [...prevFiles, ...inputFiles];
+        return allFiles;
+      }
+
+      return inputFiles;
+    });
+  }, []);
+
+  const validateFiles = (inputFiles: BladeFileList, allFiles: BladeFileList): boolean => {
+    if (accept && inputFiles.some((file) => !isFileAccepted(file, accept))) {
+      setErrorMessage(`You provided an unsupported file type. Supported file types are: ${accept}`);
+      setInternalValidationState('error');
+      return true;
+    }
+
+    if (uploadType === 'single' && inputFiles.length > 1) {
+      setErrorMessage('You can upload only one file.');
+      setInternalValidationState('error');
+      return true;
+    }
+
+    if (maxCount && allFiles.length > maxCount) {
+      setErrorMessage(`You can't upload more than ${maxCount} files.`);
+      setInternalValidationState('error');
+      return true;
+    }
+
+    if (maxSize && inputFiles.some((file) => file.size > maxSize)) {
+      setErrorMessage('File size exceeded.');
```

Comment:
do we not need to tell file size exceeded for which file?

Reply:
Thinking more on this, we don't need this info. There can be multiple files with long names too. We can add it later if requested by users.